redash/models.py

```diff
@@ -1000,6 +1006,32 @@ def outdated_queries(cls):
 
         return outdated_queries.values()
 
+    @classmethod
+    def delete_stale_resultsets(cls):
+        delete_count = 0
+        queries = Query.query.filter(Query.schedule_resultset_size != None).order_by(Query.schedule_resultset_size.desc())
+        # Multiple queries with the same text may request multiple result sets
+        # be kept. We start with the one that keeps the most, and delete both
+        # the unneeded bridge rows and result sets.
+        first_query = queries.first()
+        if first_query is not None and first_query.schedule_resultset_size:
+            resultsets = QueryResultSet.query.filter(QueryResultSet.query_rel == first_query).order_by(QueryResultSet.result_id)
+            resultset_count = resultsets.count()
+            if resultset_count > first_query.schedule_resultset_size:
+                n_to_delete = resultset_count - first_query.schedule_resultset_size
+                r_ids = [r.result_id for r in resultsets][:n_to_delete]
+                delete_count = QueryResultSet.query.filter(QueryResultSet.result_id.in_(r_ids)).delete(synchronize_session=False)
+                QueryResult.query.filter(QueryResult.id.in_(r_ids)).delete(synchronize_session=False)
```

Comment:
Is there a reason why these deletes don't contribute to the `delete_count` but the `QueryResultSet` deletes do? It seems that in `cleanup_query_results()` originally only `QueryResult` deletes were being counted so maybe these need to be counted too?

Reply:
No idea how I made that mistake. Fixed.